<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Blacklisted regions for human genome</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for hgBlacklist {seqDNAcopy}"><tr><td>hgBlacklist {seqDNAcopy}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Blacklisted regions for human genome</h2>

<h3>Description</h3>

<p>Regions in the genome that give sequencing artifacts as listed in
UCSC Genome browser track wgEncodeDacMapabilityConsensusExcludable
</p>


<h3>Usage</h3>

<pre>
  data(hgBlacklist)
</pre>


<h3>Format</h3>

<p>A list containing 25 matrices corresponding to chromosomes 1-22, X, Y &amp;
MT in that order. Each row is an interval that is to be discarded.
</p>


<h3>Details</h3>

<p>The intervals for hg19 were obtained from UCSC Genome browser track
<a href="https://genome.ucsc.edu/cgi-bin/hgFileUi?db=hg19&amp;g=wgEncodeMapability">https://genome.ucsc.edu/cgi-bin/hgFileUi?db=hg19&amp;g=wgEncodeMapability</a>
wgEncodeDacMapabilityConsensusExcludable. The intervals were padded
by 100 bases on both ends and an extra row (2^28, 2^28) added to each
chromosome for programming convenience. The hg18 and hg38 data were
obtained using liftover <a href="https://genome.ucsc.edu/cgi-bin/hgLiftOver">https://genome.ucsc.edu/cgi-bin/hgLiftOver</a>.
</p>
<p>A fragment whose mid-point lies in an interval is discarded from analysis.
</p>


<h3>References</h3>

<p><a href="https://sites.google.com/site/anshulkundaje/projects/blacklists">https://sites.google.com/site/anshulkundaje/projects/blacklists</a>
</p>

<hr /><div style="text-align: center;">[Package <em>seqDNAcopy</em> version 0.0.3 <a href="00Index.html">Index</a>]</div>
</body></html>
